client/network/src/discovery.rs

```diff
@@ -45,85 +45,131 @@
+	kademlias: HashMap<ProtocolId, Kademlia<MemoryStore>>
 }
 
-impl DiscoveryBehaviour {
-	/// Builds a new `DiscoveryBehaviour`.
-	///
-	/// `user_defined` is a list of known address for nodes that never expire.
-	pub async fn new(
-		local_public_key: PublicKey,
-		user_defined: Vec<(PeerId, Multiaddr)>,
-		enable_mdns: bool,
-		allow_private_ipv4: bool,
-		discovery_only_if_under_num: u64,
-	) -> Self {
-		if enable_mdns {
-			#[cfg(target_os = "unknown")]
-			warn!(target: "sub-libp2p", "mDNS is not available on this platform");
+impl DiscoveryConfig {
+	/// Crate a default configuration with the given public key.
+	pub fn new(local_public_key: PublicKey) -> Self {
+		DiscoveryConfig {
+			local_peer_id: local_public_key.into_peer_id(),
+			user_defined: Vec::new(),
+			allow_private_ipv4: true,
+			discovery_only_if_under_num: std::u64::MAX,
+			enable_mdns: false,
+			kademlias: HashMap::new()
 		}
+	}
 
-		let local_id = local_public_key.clone().into_peer_id();
-		let store = MemoryStore::new(local_id.clone());
-		let mut kademlia = Kademlia::new(local_id.clone(), store);
-		for (peer_id, addr) in &user_defined {
-			kademlia.add_address(peer_id, addr.clone());
+	/// Set the number of active connections at which we pause discovery.
+	pub fn discovery_limit(&mut self, limit: u64) -> &mut Self {
+		self.discovery_only_if_under_num = limit;
+		self
+	}
+
+	/// Set custom nodes which never expire, e.g. bootstrap or reserved nodes.
+	pub fn with_user_defined<I>(&mut self, user_defined: I) -> &mut Self
+	where
+		I: IntoIterator<Item = (PeerId, Multiaddr)>
+	{
+		self.user_defined.extend(user_defined);
+		self
+	}
+
+	/// Should private IPv4 addresses be reported?
+	pub fn allow_private_ipv4(&mut self, value: bool) -> &mut Self {
+		self.allow_private_ipv4 = value;
+		self
+	}
+
+	/// Should MDNS discovery be supported?
+	pub fn with_mdns(&mut self, value: bool) -> &mut Self {
+		if value && cfg!(target_os = "unknown") {
+			log::warn!(target: "sub-libp2p", "mDNS is not available on this platform")
 		}
+		self.enable_mdns = value;
+		self
+	}
 
+	/// Add discovery via Kademlia for the given protocol.
+	pub fn add_protocol(&mut self, p: ProtocolId) -> &mut Self {
+		if self.kademlias.contains_key(&p) {
+			warn!(target: "sub-libp2p", "Discovery already registered for protocol {:?}", p);
+			return self
+		}
+
+		let mut config = KademliaConfig::default();
+
+		// NB: If the protocol name derivation below is changed, check if
+		// `DiscoveryBehaviour::new_handler` is still correct.
+		let proto_name =
+			if libp2p::kad::protocol::DEFAULT_PROTO_NAME == p.as_bytes() {
+				// Temporary hack to retain backwards compatibility. Once this version
+				// has been rolled out we should remove the special handling of
+				// DEFAULT_PROTO_NAME and work only with proper protocol IDs.
+				Vec::from(p.as_bytes())
+			} else {
+				let mut v = vec![b'/'];
+				v.extend_from_slice(p.as_bytes());
+				v.extend_from_slice(b"/kad");
+				v
+			};
```

Comment:
I think it wouldn't be a bad idea to get rid of this hack.
What about something like this? (pseudocode)

```rust
impl DiscoveryConfig {
    pub fn new() {
        ...
        config.add_protocol_inner(libp2p::kad::protocol::DEFAULT_PROTO_NAME)
    }

    pub fn add_protocol(&mut self, p: ProtocolId) {
        let proto_name = "/" .. p .. "/kad";  // I know this is not Rust
        self.add_protocol_inner(proto_name)
    }

    // private
    fn add_protocol_inner(&mut self, name: Vec<u8>) { ... }
}
```